File: testPythonModule/makeHistoFile.py
import os,sys
from ROOT import *
from bjetClasses import *


def main():

    dir = 'data/'
    filesList = []
    for ifile in os.listdir(dir):
        filesList.append(ifile)
    filesList.sort()

    h_R_bl = {}

    for ifile in filesList:
        sampleID = ifile.split('.')[1]
        
        print ifile
        file = TFile.Open(dir+ifile)
        tree = file.Get("susy")

        outf = TFile.Open("histosToCompare.PythonModule.root","RECREATE")

        h_R_bl[sampleID] = TH1F("h_R_bl_"+sampleID,"",30,0,1)

        i = 0
        for entry in tree:
            if ( i % 10000 == 0):
                print '%s / %s' % (i,tree.GetEntries())
            i+=1

            t_is2b = entry.is2b
            
            if t_is2b == 0: continue

            t_jet1_pT = entry.jet1_pT
            t_jet2_pT = entry.jet2_pT
            t_jet3_pT = entry.jet3_pT
            t_jet4_pT = entry.jet4_pT            

            t_jet1_eta = entry.jet1_eta
            t_jet2_eta = entry.jet2_eta
            t_jet3_eta = entry.jet3_eta
            t_jet4_eta = entry.jet4_eta            

            t_jet1_phi = entry.jet1_phi
            t_jet2_phi = entry.jet2_phi
            t_jet3_phi = entry.jet3_phi
            t_jet4_phi = entry.jet4_phi

            t_jet1_E = entry.jet1_E
            t_jet2_E = entry.jet2_E
            t_jet3_E = entry.jet3_E
            t_jet4_E = entry.jet4_E
            
            t_jet1_MV1 = entry.jet1_MV1
            t_jet2_MV1 = entry.jet2_MV1
            t_jet3_MV1 = entry.jet3_MV1
            t_jet4_MV1 = entry.jet4_MV1

            t_lep1_pT = entry.lep1_pT
            t_lep_eta = entry.lep_eta
            t_lep_phi = entry.lep_phi
            t_lep_E = entry.lep_E

            t_mc_event_weight = entry.mc_event_weight
            t_ttbar_weight = entry.ttbar_weight
            t_Wj_sherpa_weight = entry.Wj_sherpa_weight
            
            j1 = TLorentzVector()
            j2 = TLorentzVector()
            j3 = TLorentzVector()
            j4 = TLorentzVector()
            lep = TLorentzVector()

            j1.SetPtEtaPhiE(t_jet1_pT,t_jet1_eta,t_jet1_phi,t_jet1_E)
            j2.SetPtEtaPhiE(t_jet2_pT,t_jet2_eta,t_jet2_phi,t_jet2_E)
            j3.SetPtEtaPhiE(t_jet3_pT,t_jet3_eta,t_jet3_phi,t_jet3_E)
            j4.SetPtEtaPhiE(t_jet4_pT,t_jet4_eta,t_jet4_phi,t_jet4_E)          
            lep.SetPtEtaPhiE(t_lep1_pT,t_lep_eta,t_lep_phi,t_lep_E)

            jets = bjetExtractor(j1,j2,j3,j4,t_jet1_MV1,t_jet2_MV1,t_jet3_MV1,t_jet4_MV1)
            bjetvars = bjetVariables(jets.bjet1,jets.bjet2,jets.jet1,jets.jet2,lep)
            
            eventWeight = t_mc_event_weight
            if t_ttbar_weight != -999:
                eventWeight *= t_ttbar_weight
            if t_Wj_sherpa_weight != -999:
                eventWeight *= t_Wj_sherpa_weight
            
            
            h_R_bl[sampleID].Fill(bjetvars.getR_bl(),eventWeight)
            
 

    #Don't know why, but have to use this method to get it written into the file
    outf.cd()
    for sampleID in list(h_R_bl.keys()):
        h_R_bl[sampleID].Write()
    

if __name__ == "__main__":
    main()

File: testCppModule/makeHistoFile.cxx
#include <iostream>
#include "TLorentzVector.h"
#include "TH1.h"
#include "TFile.h"
#include "TTree.h"

int makeHistoFile(TString fileName, TString sampleID){

  gSystem->Load("bjetExtractor_cxx.so");
  gSystem->Load("bjetVariables_cxx.so");

  TFile *f = TFile::Open(fileName);
  TTree *tree = f->Get("susy");

  TFile *outf = TFile::Open("histosToCompare_"+sampleID+".root","RECREATE");

  int is2b;
  float lep1_pT,jet1_pT,jet2_pT,jet3_pT,jet4_pT;
  float lep_eta,jet1_eta,jet2_eta,jet3_eta,jet4_eta;
  float lep_phi,jet1_phi,jet2_phi,jet3_phi,jet4_phi;
  float lep_E,jet1_E,jet2_E,jet3_E,jet4_E;
  float jet1_MV1,jet2_MV1,jet3_MV1,jet4_MV1;

  double mc_event_weight;
  float ttbar_weight,Wj_sherpa_weight;
  
  tree->SetBranchAddress("lep1_pT",&lep1_pT);
  tree->SetBranchAddress("jet1_pT",&jet1_pT);
  tree->SetBranchAddress("jet2_pT",&jet2_pT);
  tree->SetBranchAddress("jet3_pT",&jet3_pT);
  tree->SetBranchAddress("jet4_pT",&jet4_pT);

  tree->SetBranchAddress("lep_eta",&lep_eta);
  tree->SetBranchAddress("jet1_eta",&jet1_eta);
  tree->SetBranchAddress("jet2_eta",&jet2_eta);
  tree->SetBranchAddress("jet3_eta",&jet3_eta);
  tree->SetBranchAddress("jet4_eta",&jet4_eta);

  tree->SetBranchAddress("lep_phi",&lep_phi);
  tree->SetBranchAddress("jet1_phi",&jet1_phi);
  tree->SetBranchAddress("jet2_phi",&jet2_phi);
  tree->SetBranchAddress("jet3_phi",&jet3_phi);
  tree->SetBranchAddress("jet4_phi",&jet4_phi);

  tree->SetBranchAddress("lep_E",&lep_E);
  tree->SetBranchAddress("jet1_E",&jet1_E);
  tree->SetBranchAddress("jet2_E",&jet2_E);
  tree->SetBranchAddress("jet3_E",&jet3_E);
  tree->SetBranchAddress("jet4_E",&jet4_E);

  tree->SetBranchAddress("jet1_MV1",&jet1_MV1);
  tree->SetBranchAddress("jet2_MV1",&jet2_MV1);
  tree->SetBranchAddress("jet3_MV1",&jet3_MV1);
  tree->SetBranchAddress("jet4_MV1",&jet4_MV1);

  tree->SetBranchAddress("is2b",&is2b);
  
  tree->SetBranchAddress("mc_event_weight",&mc_event_weight);
  tree->SetBranchAddress("ttbar_weight",&ttbar_weight);
  tree->SetBranchAddress("Wj_sherpa_weight",&Wj_sherpa_weight);


  int N = tree->GetEntries();
  TH1F *h_R_bl = new TH1F("h_R_bl_"+sampleID,"R_bl;R_bl;",30,0,1);
  
  std::cout << "Looping over file " << f->GetName() << " and TTree " << tree->GetName() << std::endl;
  std::cout << "Total number of events: " << N << std::endl;

  int step;
  if( N < 1000 ){ step = 100; }
  else if( N < 10000){ step = 1000;}
  else if( N < 100000) { step = 10000; }
  else { step = 100000; }

  for(int i = 0; i < N; i++){
    if (i % step == 0){
      if(i != 0){
	std::cout << "Event " << i << " / " << N << std::endl;
      }
    }
    
    tree->GetEntry(i);
    
    if (is2b == 0) continue;

    TLorentzVector j1,j2,j3,j4,lep;
    j1.SetPtEtaPhiE(jet1_pT,jet1_eta,jet1_phi,jet1_E);
    j2.SetPtEtaPhiE(jet2_pT,jet2_eta,jet2_phi,jet2_E);
    j3.SetPtEtaPhiE(jet3_pT,jet3_eta,jet3_phi,jet3_E);
    j4.SetPtEtaPhiE(jet4_pT,jet4_eta,jet4_phi,jet4_E);
    lep.SetPtEtaPhiE(lep1_pT,lep_eta,lep_phi,lep_E);

    bjetExtractor jets(j1,j2,j3,j4,jet1_MV1,jet2_MV1,jet3_MV1,jet4_MV1);
    bjetVariables bjetvars(jets.bjet1,jets.bjet2,jets.jet1,jets.jet2,lep);
    
    float eventWeight = mc_event_weight;
    if(ttbar_weight != -999){
      eventWeight*=ttbar_weight;
    }
    if(Wj_sherpa_weight != -999){
      eventWeight*=Wj_Sherpa_weight;
    }
    
    h_R_bl->Fill(bjetvars.getR_bl(),eventWeight);
    
  }

  outf->cd();
  h_R_bl->Write();
  
  h_R_bl->Draw();
  
  return 0;
}

